index.ts

```diff
@@ -70,61 +72,91 @@ function appendLog(payload: any) {
   );
 }
 
-let migrated = 0;
-let alreadyExists = 0;
-
-async function processUserToClerk(userData: User) {
+export async function processUserToClerk(userData: User) {
   try {
     const parsedUserData = userSchema.safeParse(userData);
     if (!parsedUserData.success) {
-      throw parsedUserData.error;
+      return { error: parsedUserData.error, type: "validation" } as const;
     }
-    await createUser(parsedUserData.data);
 
-    migrated++;
+    return { data: await createUser(parsedUserData.data) } as const;
   } catch (error) {
     if (error.status === 422) {
-      appendLog({ userId: userData.userId, ...error });
-      alreadyExists++;
-      return;
+      return { error, type: "already_exists" } as const;
     }
 
-    // Keep cooldown in case rate limit is reached as a fallback if the thread blocking fails
     if (error.status === 429) {
-      console.log(`Waiting for rate limit to reset`);
-      await cooldown();
-
-      console.log("Retrying");
-      // conditional recursion
-      return processUserToClerk(userData);
+      return { error, type: "rate_limit" } as const;
     }
 
-    appendLog({ userId: userData.userId, ...error });
+    return { error, type: "unknown" } as const;
```

Comment:
This function is only responsible for making the request and returning a status so that the caller can handle individual cases.